cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -428,8 +428,11 @@ private void writeToHeartBeatDataset(@Nullable RunRecordMeta recordedRunRecord,
           }
         });
       case DEPROVISIONING:
-        appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
-        return Optional.of(provisioningService.deprovision(programRunId, datasetContext));
+        RunRecordMeta recordedMeta = appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
+        if (recordedMeta != null) {
```

Comment:
can you add a comment here too about why you are doing a null check?
Remainder LGTM